File: tests/mat-mul/Python/src/matrix.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# standard library
import random
import sys
import threading

# 3rd party packages

# local source

# --------------------------------------------------------------------------- #
__author__ = "Juan Luis Padilla Salomé"
__copyright__ = "Copyright 2021"
__credits__ = ["University of Cantabria", "Pablo Abad", "Pablo Prieto"]
__license__ = "GPL"
__version__ = "1.0.0"
__maintainer__ = "Juan Luis Padilla Salomé"
__email__ = "juan-luis.padilla@alumnos.unican.es"
__status__ = "Production"
# --------------------------------------------------------------------------- #

# --------------------------------------------------------------------------- #


class matrix(object):
    """
    A class used to represent a Matrix and perform different operations with
    matrices of different dimensions.

    Attributes
    ----------
    rows_default : int
        default value of rows for generate a matrix
    cols_default : int
        default value of cols for generate a matrix

    Methods
    -------
    init_empty(rows=None, cols=None)
        desc
    init_random(rows=None, cols=None)
        desc
    init_seq(rows=None, cols=None)
        desc
    init_zeros(rows=None, cols=None)
        desc
    mat_mul(M_a=None, M_b=None)
        desc
    """

    # Attributes
    # self.MAX_RANDOM
    # self.NUM_THREADS

    def __init__(self):
        """Constructor."""

        super(matrix, self).__init__()

        # Attributes
        self.MAX_RANDOM = 10
        self.NUM_THREADS = 16

        # Establish the warning format
        # warnings.formatwarning = self.__warning_on_one_line

    # -------------------------------------------------------------------- #

    def print_matrix(self, M):
        """Returns a Numpy matrix initialized with the function empty().

        If the arguments `rows` or `cols` aren't passed in, the default
        attributes are used.

        Parameters
        ----------
        rows : int, optional
            The number of rows of the matrix (default is None)
        
        cols : int, optional
            The number of cols of the matrix (default is None)

        Returns
        -------
        numpy.matrix
            a floating point matrix of size rows x cols.

        Raises
        ------
        ValueError
            If the arguments passed in are not positive.
        """

        rows = len(M)
        cols = len(M[-1])

        mat_str = "\n"
        for i in range(rows):
            mat_str += "|\t"
            for j in range(cols):
                mat_str += str(M[i][j]) + "\t"
                if j == cols - 1:
                    mat_str += "|\n"
        print(mat_str)
    # -------------------------------------------------------------------- #

    def init_rand(self, rows, cols):
        """Returns a Numpy matrix with its values randomly initialized.

        If the arguments `rows` or `cols` aren't passed in, the default
        attributes are used.

        Parameters
        ----------
        rows : int, optional
            The number of rows of the matrix (default is None)
        
        cols : int, optional
            The number of cols of the matrix (default is None)

        Returns
        -------
        numpy.matrix
            a floating point matrix of size rows x cols.

        Raises
        ------
        ValueError
            If the arguments passed in are not positive.
        """

        M = []
        for r in range(rows):
            arr = []
            for c in range(cols):
                arr.append(float(random.randrange(self.MAX_RANDOM)))
            M.append(arr)

        return M
    # -------------------------------------------------------------------- #

    def init_seq(self, rows, cols):
        """Returns a Numpy matrix with its values sequentially initialized.

        If the arguments `rows` or `cols` aren't passed in, the default
        attributes are used.

        Parameters
        ----------
        rows : int, optional
            The number of rows of the matrix (default is None)
        
        cols : int, optional
            The number of cols of the matrix (default is None)

        Returns
        -------
        matrix
            a floating point matrix of size rows x cols.

        """

        M = []
        for r in range(rows):
            arr = []
            for c in range(cols):
                arr.append(float(r * cols + c))
            M.append(arr)

        return M
    # -------------------------------------------------------------------- #

    def mat_mul(self, M_a, M_b):
        """Returns a Numpy matrix product of a multiplication of matrices.

        If the arguments `M_a` or `M_a` aren't passed in, the operation
        cannot be performed.

        Parameters
        ----------
        M_a : numpy.matrix
            A matrix, first term of the multiplication (default is None)
        
        M_b : numpy.matrix
            A matrix, second term of the multiplication (default is None)

        Returns
        -------
        numpy.matrix
            a floating point matrix.

        Raises
        ------
        MatricesUndefinedError
            If there is no argument passed.
        """

        rows_a = len(M_a)
        cols_a = len(M_a[-1])
        rows_b = len(M_b)
        cols_b = len(M_b[-1])

        if cols_a != rows_b:
            print("[ERROR] #columns A must be equal to #rows B.\n")
            sys.exit(-1)

        M_c = []
        for i in range(rows_a):
            arr = []
            for k in range(cols_b):
                sum = 0.0
                for j in range(cols_a):
                    sum += float(M_a[i][j]) * float(M_b[j][k])
                arr.append(sum)
            M_c.append(arr)
        return M_c
    # -------------------------------------------------------------------- #

    def mat_mul_transpose(self, M_a, M_b):
        """Returns a Numpy matrix product of a multiplication of matrices.

        If the arguments `M_a` or `M_a` aren't passed in, the operation
        cannot be performed.

        Parameters
        ----------
        M_a : numpy.matrix
            A matrix, first term of the multiplication (default is None)
        
        M_b : numpy.matrix
            A matrix, second term of the multiplication (default is None)

        Returns
        -------
        numpy.matrix
            a floating point matrix.

        Raises
        ------
        MatricesUndefinedError
            If there is no argument passed.
        """

        rows_a = len(M_a)
        cols_a = len(M_a[-1])
        rows_b = len(M_b)
        cols_b = len(M_b[-1])

        if cols_a != rows_b:
            print("[ERROR] #columns A must be equal to #rows B.\n")
            sys.exit(-1)

        # Calculate the transpose
        M_b_Tr = []
        for i in range(rows_b):
            arr = []
            for j in range(cols_b):
                arr.append(M_b[j][i])
            M_b_Tr.append(arr)

        M_c = []
        for i in range(rows_a):
            arr = []
            for k in range(cols_b):
                sum = 0.0
                for j in range(cols_a):
                    sum += float(M_a[i][j]) * float(M_b_Tr[k][j])
                arr.append(sum)
            M_c.append(arr)
        return M_c
    # -------------------------------------------------------------------- #

    def __multi(self, M_a, M_b, M_c, cols_c, rows_c):
        cols_a = len(M_a[-1])

        aux = []
        for i in range(rows_c[0], rows_c[1]):
            arr = []
            for k in range(cols_c[0], cols_c[1]):
                sum = 0.0
                for j in range(cols_a):
                    sum += float(M_a[i][j]) * float(M_b[j][k])
                arr.append(sum)
            aux.append(arr)

        # Store the values in the result matrix
        i = 0
        for j in range(rows_c[0], rows_c[1]):
            M_c[j] = aux[i]
            i += 1
    # -------------------------------------------------------------------- #

    def mat_mul_multithread(self, M_a, M_b):

        # NUM_THREADS = 2
        rows_a = len(M_a)
        cols_a = len(M_a[-1])
        rows_b = len(M_b)
        cols_b = len(M_b[-1])

        if cols_a != rows_b:
            print("[ERROR] #columns A must be equal to #rows B.\n")
            sys.exit(-1)

        rows_per_thread = int(rows_a / self.NUM_THREADS)
        rest_of_matrix = rows_a % self.NUM_THREADS

        M_c = [[0] * cols_b] * rows_a

        # Create and start the threads
        threads = []
        for i in range(self.NUM_THREADS):
            # Calculate the params to pass them to the thread
            # TODO: The last thread operates the rest. Modify in a future and
            # TODO: let the first thread to end, operate the rest.
            cols_c_start = 0
            cols_c_end = cols_b
            rows_c_start = rows_per_thread * i
            rows_c_end = rows_c_start + rows_per_thread

            if (i == self.NUM_THREADS - 1) and (rest_of_matrix != 0):
                rows_c_end += rest_of_matrix

            cols_c = [cols_c_start, cols_c_end]
            rows_c = [rows_c_start, rows_c_end]

            t = threading.Thread(target=self.__multi, args=(M_a, M_b, M_c, cols_c, rows_c,))
            threads.append(t)
            t.start()

        # Join the threads
        for t in threads:
            t.join()

        return M_c
    # -------------------------------------------------------------------- #


    # -------------------------------------------------------------------- #
    # define Python user-defined exceptions

    class Error(Exception):
        """Base class for other exceptions"""
        pass

    class MatricesUndefinedError(Error):
        """Raised when the input values are less than two arguments."""
        pass
    # -------------------------------------------------------------------- #
# ------------------------------------------------------------------------ #

File: tests/mat-mul/C/src/matrix.c
#include <pthread.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "matrix.h"

MAT_MUL *mat_multiplying;

char *arr_to_str(const double *M,
                 const unsigned rows, const unsigned cols)
{
    size_t i, j;
    char *str, *aux;
    int max_digit_in_num = 18; // 18 chars/number
    str = (char *)malloc(max_digit_in_num * sizeof(char) * rows * cols);
    aux = (char *)malloc(max_digit_in_num * sizeof(char));
    strcpy(str, "\n");
    for (i = 0; i < rows; i++)
    {
        strcat(str, "|\t");
        for (j = 0; j < cols; j++)
        {
            sprintf(aux, "%8.3f", M[i * cols + j]);
            if (j < cols - 1)
            {
                strcat(aux, "\t");
            }
            else if (j < cols)
            {
                strcat(aux, "\t|\n");
            }
            strcat(str, aux);
        }
    }
    free(aux);
    return str;
}

void init_rand(double *M, const unsigned rows, const unsigned cols)
{
    size_t i, j;
    for (i = 0; i < rows; i++)
    {
        for (j = 0; j < cols; j++)
        {
            M[i * cols + j] = (double)(rand()) /
                              ((double)RAND_MAX) * MAX_RANDOM;
        }
    }
}

void init_seq(double *M, const unsigned rows, const unsigned cols)
{
    size_t i, j;
    for (i = 0; i < rows; i++)
    {
        for (j = 0; j < cols; j++)
        {
            M[i * cols + j] = i * cols + j;
        }
    }
}

void mat_free(double *M)
{
    free(M);
}

double *mat_mul(const double *M_a,
                const unsigned rows_a, const unsigned cols_a,
                const double *M_b,
                const unsigned rows_b, const unsigned cols_b)
{
    size_t i, j, k;
    if (cols_a != rows_b)
    {
        fprintf(stderr, "[ERROR] #columns A must be equal to #rows B.\n");
        exit(EXIT_FAILURE);
    }
    double *M_c = (double *)malloc(rows_a * cols_b * sizeof(double));
    if (M_c == NULL)
    {
        fprintf(stderr, "[ERROR] Couldn't allocate memory.\n");
        exit(EXIT_FAILURE);
    }
    for (i = 0; i < rows_a; i++)
    {
        for (k = 0; k < cols_b; k++)
        {
            double sum = 0.0;
            for (j = 0; j < cols_a; j++)
            {
                sum += M_a[i * cols_a + j] * M_b[j * cols_b + k];
            }
            M_c[i * cols_b + k] = sum;
        }
    }
    return M_c;
}

double *mat_mul_multithread(const double *M_a,
                            const unsigned rows_a, const unsigned cols_a,
                            const double *M_b,
                            const unsigned rows_b, const unsigned cols_b)
{
    // TODO: Let's start with square matrices (nxn)
    size_t i;
    pthread_attr_t attr;
    pthread_t threads[NUM_THREADS];
    MAT_MUL_ARG args[NUM_THREADS];

    if (cols_a != rows_b)
    {
        fprintf(stderr, "[ERROR] #columns A must be equal to #rows B.\n");
        exit(EXIT_FAILURE);
    }
    double *M_c = (double *)malloc(rows_a * cols_b * sizeof(double));
    if (M_c == NULL)
    {
        fprintf(stderr, "[ERROR] Couldn't allocate memory.\n");
        exit(EXIT_FAILURE);
    }

    // Creates a struct with all the data of the matrices
    mat_multiplying = (MAT_MUL *)malloc(sizeof(MAT_MUL));
    mat_multiplying->M_a = (double *)M_a;
    mat_multiplying->rows_a = rows_a;
    mat_multiplying->cols_a = cols_a;
    mat_multiplying->M_b = (double *)M_b;
    mat_multiplying->rows_b = rows_b;
    mat_multiplying->cols_b = cols_b;

    // TODO: Let's make it ez and let the threads operate all the row:
    unsigned int rows_per_thread = rows_a / NUM_THREADS;
    unsigned int rest_of_matrix = rows_a % NUM_THREADS;

    // initializes the thread attributes with default values
    pthread_attr_init(&attr);

    // Creating NUM_THREADS threads, each evaluating its own part
    for (i = 0; i < NUM_THREADS; i++)
    {
        // Creating the struct to pass with all the parameters needed
        args[i].id = i;
        args[i].M_c = M_c;
        // TODO: The last thread operates the rest. Modify in a future and
        // TODO: let the first thread to end, operate the rest.
        args[i].cols_c_start = 0;
        args[i].cols_c_end = cols_b;
        args[i].rows_c_start = rows_per_thread * i;
        args[i].rows_c_end = args[i].rows_c_start + rows_per_thread;

        if (i == NUM_THREADS - 1 && rest_of_matrix != 0)
        {
            args[i].rows_c_end += rest_of_matrix;
        }

        // printf("[PID=%d] COLS[%d - %d] ROWS[%d - %d]\n", args[i].id,
        //        args[i].cols_c_start, args[i].cols_c_end,
        //        args[i].rows_c_start, args[i].rows_c_end);

        if ((pthread_create(&threads[i], &attr, (void *)__multi,
                            (void *)&args[i])) < 0)
        {
            fprintf(stderr, "[Error] Couldn't create thread #%ld.\n", i);
            exit(EXIT_FAILURE);
        }
    }
    // Joining and waiting for all threads to complete
    for (i = 0; i < NUM_THREADS; i++)
    {
        pthread_join(threads[i], NULL);
    }
    free(mat_multiplying);
    pthread_attr_destroy(&attr);
    return M_c;
}

void *__multi(void *arg)
{
    size_t i, j, k;
    MAT_MUL_ARG *aux = (MAT_MUL_ARG *)arg;

    double *M_a = mat_multiplying->M_a; // Matrix A
    // unsigned rows_a = mat_multiplying->rows_a; // Number of rows in matrix A
    unsigned cols_a = mat_multiplying->cols_a; // Number of cols in matrix A
    double *M_b = mat_multiplying->M_b;        // Matrix B
    // unsigned rows_b = mat_multiplying->rows_b; // Number of rows in matrix B
    unsigned cols_b = mat_multiplying->cols_b; // Number of cols in matrix B

    for (i = aux->rows_c_start; i < aux->rows_c_end; i++)
    {
        for (k = aux->cols_c_start; k < aux->cols_c_end; k++)
        {
            double sum = 0.0;
            for (j = 0; j < cols_a; j++)
            {
                sum += M_a[i * cols_a + j] * M_b[j * cols_b + k];
            }
            aux->M_c[i * cols_b + k] = sum;
        }
    }
    pthread_exit(NULL);
    free(aux);
}

double *mat_mul_transpose(const double *M_a,
                          const unsigned rows_a, const unsigned cols_a,
                          const double *M_b,
                          const unsigned rows_b, const unsigned cols_b)
{
    size_t i, j, k;
    if (cols_a != rows_b)
    {
        fprintf(stderr, "[ERROR] #columns A must be equal to #rows B.\n");
        exit(EXIT_FAILURE);
    }
    double *M_b_Tr = (double *)malloc(rows_b * cols_b * sizeof(double));
    double *M_c = (double *)malloc(rows_a * cols_b * sizeof(double));
    if (M_c == NULL || M_b_Tr == NULL)
    {
        fprintf(stderr, "[ERROR] Couldn't allocate memory.\n");
        exit(EXIT_FAILURE);
    }
    // Calculate the transpose
    for (i = 0; i < rows_b; i++)
    {
        for (j = 0; j < cols_b; j++)
        {
            M_b_Tr[j * rows_b + i] = M_b[i * cols_b + j];
        }
    }
    for (i = 0; i < rows_a; i++)
    {
        for (k = 0; k < cols_b; k++)
        {
            double sum = 0.0;
            for (j = 0; j < cols_a; j++)
            {
                sum += M_a[i * cols_a + j] * M_b_Tr[k * rows_b + j];
            }
            M_c[i * cols_b + k] = sum;
        }
    }
    free(M_b_Tr);
    return M_c;
}
